sw/device/lib/testing/usb_testutils.c

```diff
@@ -14,54 +14,134 @@
+  }
+  //  Is this the last packet?
+  uint32_t next_offset = buf->offset + bytes_written;
+  *last = true;
+  if (bytes_written == max_packet) {
+    if (next_offset < buf->length || (buf->flags & kUsbTestutilsXfrEmployZLP)) {
+      *last = false;
+    }
+  } else {
+    CHECK(bytes_written < max_packet);
+  }
+
+  buf->offset = next_offset;
+  return true;
+}
+
+// Internal function to perform the next part of a larger buffer transfer
+static bool usb_testutils_buffer_part(usb_testutils_ctx_t *ctx, uint8_t ep,
+                                      usb_testutils_buffer_t *buf) {
+  // Do we need to prepare a packet?
+  if (!buf->next_valid &&
+      !usb_testutils_part_prepare(ctx, buf, &buf->next_part, &buf->last)) {
+    return false;
+  }
+
+  // Send the existing prepared packet
+  CHECK_DIF_OK(dif_usbdev_send(ctx->dev, ep, &buf->next_part));
+  buf->next_valid = false;
+
+  // If we're double-buffering, request and fill another buffer immediately;
+  // we'll then be able to supply it much more promptly later...
+  if ((buf->flags & kUsbTestutilsXfrDoubleBuffered) && !buf->last) {
+    buf->next_valid =
+        usb_testutils_part_prepare(ctx, buf, &buf->next_part, &buf->last);
+  }
+
+  return true;
+}
+
 void usb_testutils_poll(usb_testutils_ctx_t *ctx) {
   uint32_t istate;
 
   // Collect a set of interrupts
   CHECK_DIF_OK(dif_usbdev_irq_get_state(ctx->dev, &istate));
 
   if (!istate) {
-    // Nothing new to do right now, but keep buffers available for reception
-    CHECK_DIF_OK(dif_usbdev_fill_available_fifo(ctx->dev, ctx->buffer_pool));
     return;
   }
 
-  // Record bus frame
-  if ((istate & (1u << kDifUsbdevIrqFrame))) {
-    // The first bus frame is 1
-    CHECK_DIF_OK(dif_usbdev_status_get_frame(ctx->dev, &ctx->frame));
-    ctx->got_frame = true;
-  }
-
   // Process IN completions first so we get the fact that send completed
   // before processing a response to that transmission
   // This is also important for device IN performance
   if (istate & (1u << kDifUsbdevIrqPktSent)) {
     uint16_t sentep;
     CHECK_DIF_OK(dif_usbdev_get_tx_sent(ctx->dev, &sentep));
     TRC_C('a' + sentep);
-    for (unsigned ep = 0; ep < USBDEV_NUM_ENDPOINTS; ep++) {
-      if (sentep & (1 << ep)) {
+    unsigned ep = 0u;
+    while (sentep && ep < USBDEV_NUM_ENDPOINTS) {
+      if (sentep & (1u << ep)) {
         // Free up the buffer and optionally callback
         CHECK_DIF_OK(
             dif_usbdev_clear_tx_status(ctx->dev, ctx->buffer_pool, ep));
 
-        if (ctx->tx_done_callback[ep]) {
-          ctx->tx_done_callback[ep](ctx->ep_ctx[ep]);
+        // If we have a larger transfer in progress, continue with that
+        usb_testutils_buffer_t *buf = &ctx->in[ep].buf;
+        usb_testutils_xfr_result_t res = kUsbTestutilsXfrResultOk;
+        bool done = true;
+        if (buf->buffer) {
+          if (buf->next_valid || !buf->last) {
+            if (usb_testutils_buffer_part(ctx, ep, buf)) {
```

Comment:
Is there a more descriptive name for this function? It's really unclear what sort of conditional check comes from a "buffer part".